std/digest/xxh.d

```diff
@@ -0,0 +1,3177 @@
+    0x8f, 0xf8, 0xb8, 0xd1, 0x7a, 0xd0, 0x31, 0xce, 0x45, 0xcb, 0x3a, 0x8f,
+    0x95, 0x16, 0x04, 0x28, 0xaf, 0xd7, 0xfb, 0xca, 0xbb, 0x4b, 0x40, 0x7e,
+];
+
+/*
+ * Downcast + upcast is usually better than masking on older compilers like
+ * GCC 4.2 (especially 32-bit ones), all without affecting newer compilers.
+ *
+ * The other method, (x & 0xFFFFFFFF) * (y & 0xFFFFFFFF), will AND both operands
+ * and perform a full 64x64 multiply -- entirely redundant on 32-bit.
+ */
+private ulong XXH_mult32to64(uint x, uint y) @safe pure nothrow @nogc
+{
+    return (cast(ulong)(x) * cast(ulong)(y));
+}
+
+/** Calculates a 64.128-bit long multiply.
+ *
+ * Uses `__uint128_t` and `_umul128` if available, otherwise uses a scalar
+ * version.
+ *
+ * Param: lhs , rhs The 64-bit integers to be multiplied
+ * Return: The 128-bit result represented in an XXH128_hash_t.
+ */
+private XXH128_hash_t XXH_mult64to128(ulong lhs, ulong rhs) @safe pure nothrow @nogc
+{
+    static if (is(ucent))
+    {
+        //import std.int128;
+        //alias uint128_t = Int128;
+
+        const uint128_t product = cast(uint128_t_) lhs * cast(uint128_t_) rhs;
+        XXH128_hash_t r128;
+        // r128.low64  = cast(ulong) (product);
+        // r128.high64 = cast(ulong) (product >> 64);
+        r128.low64 = product.data.lo;
+        r128.high64 = product.data.hi;
+    }
+    else
+    {
+        /* First calculate all of the cross products. */
+        const ulong lo_lo = XXH_mult32to64(lhs & 0xFFFFFFFF, rhs & 0xFFFFFFFF);
+        const ulong hi_lo = XXH_mult32to64(lhs >> 32, rhs & 0xFFFFFFFF);
+        const ulong lo_hi = XXH_mult32to64(lhs & 0xFFFFFFFF, rhs >> 32);
+        const ulong hi_hi = XXH_mult32to64(lhs >> 32, rhs >> 32);
+
+        /* Now add the products together. These will never overflow. */
+        const ulong cross = (lo_lo >> 32) + (hi_lo & 0xFFFFFFFF) + lo_hi;
+        const ulong upper = (hi_lo >> 32) + (cross >> 32) + hi_hi;
+        const ulong lower = (cross << 32) | (lo_lo & 0xFFFFFFFF);
+
+        XXH128_hash_t r128;
+        r128.low64 = lower;
+        r128.high64 = upper;
+
+    }
+    return r128;
+}
+
+/** Calculates a 64-bit to 128-bit multiply, then XOR folds it.
+ *
+ * The reason for the separate function is to prevent passing too many structs
+ * around by value. This will hopefully inline the multiply, but we don't force it.
+ *
+ * Param: lhs , rhs The 64-bit integers to multiply
+ * Return: The low 64 bits of the product XOR'd by the high 64 bits.
+ * See: XXH_mult64to128()
+ */
+private ulong XXH3_mul128_fold64(ulong lhs, ulong rhs) @safe pure nothrow @nogc
+{
+    XXH128_hash_t product = XXH_mult64to128(lhs, rhs);
+    return product.low64 ^ product.high64;
+}
+
+/* Seems to produce slightly better code on GCC for some reason. */
+pragma(inline, true)
+private ulong XXH_xorshift64(ulong v64, int shift) @safe pure nothrow @nogc
+{
+    assert(0 <= shift && shift < 64, "shift out of range");
```

Comment:
This would be the contract programming approach.

```suggestion
private ulong XXH_xorshift64(ulong v64, int shift) @safe pure nothrow @nogc
    in(0 <= shift && shift < 64, "shift out of range")
{
```

Reply:
Ok. I tried contract programming with `in { ... }`, which results in unreadable code. It also did not work for all places. So I stashed these changes. Using `in(...)` looks much better, and works for all these asserts at the beginning of a function.

Reply:
Changed.